### DISCUSSION TITLE
Support for Per-Window Theme Customization

### DISCUSSION THREAD

**Discussion by KMJ-007** [zed-industries/zed#24010] [Feature Requests]
Currently, Zed applies themes globally across all windows, which limits users who wish to visually distinguish between different projects or environments by applying unique themes to individual windows or provide theme API to achieve this so community can build extension, or store theme setting at project level in .zed folder, etc..

### Zed Version and System Specs

Zed: v0.166.1 (Zed)
OS: macOS 15.1.1
Memory: 8 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help present your vision of the feature

vscode has this extension: https://marketplace.visualstudio.com/items?itemName=stuart.unique-window-colors

**Comment by msageryd** [comment]
I was just about to post a similar feature request. To me it would be enough if theme_overides or even part of it would be used per project. I'll add my post here instead of opening a new:

## Summary

Add support for theme_overrides in project/local settings files so appearance can be customized per project (e.g., window/title bar, background, accent colors). This enables quick visual differentiation between multiple open projects without creating full themes or switching manually.

## Motivation / use case

- I’m migrating from VSCode. Each workspace there has a different window/title bar color, which makes it much easier to distinguish between many open projects at a glance.

- In Zed, local/project settings intentionally exclude appearance (docs: configuring-zed), so I can’t set theme or theme_overrides per project. Manually switching themes per window works but doesn’t scale when juggling many repos.

- I don’t want to maintain separate theme files for every project or publish extensions; a few color tweaks per repo would be enough.

## What I’m asking for

- Allow theme_overrides in local/project settings.json.

- At minimum, permit a small, well-defined subset of appearance keys that are safe per project, such as:
	- title_bar.background or tab_bar.background (window identity color)

	- panel.background, accent, border

	- optionally editor.background (for project identity)


- Keep theme selection itself at user scope if desired; only the overrides would be project-scoped.

### Proposed precedence and merge behavior

- Precedence: project theme_overrides > user theme_overrides > base theme.

- Merge: shallow merge by key; unknown keys ignored. If a project override sets just title_bar.background, it should not affect other user-level styling.

- Scope: apply per-window based on the project root of that window.

### Why not local themes?

- Local themes work, but:

- They require maintaining full theme files (verbose, brittle to schema changes).

- Switching is manual and not tied to the project a window has opened.

- For my use case, I only need a few color hints per repo to differentiate windows.

### Benefits

- Better multi-repo ergonomics without heavy customization.

- Encourages teams to keep brand or environment-specific visual cues in-repo.

- Avoids proliferation of nearly identical theme files.

### Compatibility and safety

- Default behavior unchanged if a project has no theme_overrides.

- Strictly validate keys; ignore anything outside the allowed subset if you want to keep appearance surface small at project scope.

- Consider a window badge color as a minimal first step if broader overrides feel too open-ended.

### Example

User settings (global):
```
{
  "theme": "Monokai-og",
  "theme_overrides": {
    "selection.background": { "color": "#57526888" }
  }
}
```

Project .zed/settings.json:
```
{
  "theme_overrides": {
    "title_bar.background": { "color": "#134E4A" },
    "tab_bar.background": { "color": "#0F3E39" } 
  }
}
```

Result: project window has a teal-tinted chrome while retaining the user’s base theme and global selection color.

### Links

- Docs note about local settings excluding appearance: https://zed.dev/docs/configuring-zed

- Request is specifically for enabling theme_overrides at project scope with clear precedence.

**Reply by msageryd** [reply]
After some digging I found more duplicates for this. Sorry for posting the above duplicate. Here are the others I found:
https://github.com/zed-industries/zed/discussions/32293
https://github.com/zed-industries/zed/issues/13300

**Comment by behavio1** [comment]
Genuinely frustrating when you have multiple projects open simultaneously — every window is identical. What's everyone's current workaround?

**Reply by gernotstarke** [reply]
stick with VS-Code for most of my daily work...

**Comment by nigelgrange** [comment]
This is the only thing stopping me from switching from VSCode to Zed - I have at least 5 projects open at any time, and I need to visually differentiate which window belongs to which project.

**Reply by Nkr1shna** [reply]
Not sure if this is useful, but the current multi workspace in a single window setup works well for me. I use cmd+opt+o to switch between open projects.

**Comment by CaptainCodeAU** [comment]
> Currently, Zed applies themes globally across all windows, which limits users who wish to visually distinguish between different projects or environments by applying unique themes to individual windows or provide theme API to achieve this so community can build extension, or store theme setting at project level in .zed folder, etc..
> 
> ### Zed Version and System Specs
> Zed: v0.166.1 (Zed) OS: macOS 15.1.1 Memory: 8 GiB Architecture: aarch64
> 
> ### If applicable, add mockups / screenshots to help present your vision of the feature
> vscode has this extension: https://marketplace.visualstudio.com/items?itemName=stuart.unique-window-colors


People will say run separate instances with --user-data-dir. don't bother. that flag moves your data dir but not your settings reliably (there's a known bug where the CLI just loads your normal settings anyway), and even when it works you've now got a completely separate Zed per project. separate extensions, separate login, separate everything. that's not a window color, that's maintaining 5 parallel installs. way too much friction to actually keep up.

The reason I care: I'm dyslexic and have ADHD and I run 4-5 projects open at once all day. every window is identical so I constantly tab into the wrong one and start editing the wrong repo. color is the one thing my brain actually locks onto, text labels in the title bar I just slide right past. @msageryd 's proposal above is exactly right, even just title_bar.background per project would fix this for me. it's an accessibility thing, not a cosmetic one.

**Comment by 42piratas** [comment]
https://github.com/zed-industries/zed/pull/58755
